source/develop/release-management/features/storage/cd-block-storage.md

```diff
@@ -0,0 +1,228 @@
+                <ovirt-vm:leaseOffset type="int">105906176</ovirt-vm:leaseOffset>
+                <ovirt-vm:leasePath>/dev/88252cf6-381e-48f0-8795-a294a32c7149/leases</ovirt-vm:leasePath>
+                <ovirt-vm:path>/rhev/data-center/mnt/blockSD/88252cf6-381e-48f0-8795-a294a32c7149/images/89f05c7d-b961-4935-993f-514499024515/626a493f-5214-4337-b580-96a1ce702c2a</ovirt-vm:path>
+                <ovirt-vm:volumeID>626a493f-5214-4337-b580-96a1ce702c2a</ovirt-vm:volumeID>
+            </ovirt-vm:volumeChainNode>
+        </ovirt-vm:volumeChain>
+    </ovirt-vm:device>
+```
+
+When CD is not loaded, this element is still present (as CD-ROM is always present), but is empty, e.g.:
+
+```
+    <ovirt-vm:device devtype="disk" name="sdc"/>
+```
+
+### Bugs
+
+- [Bug 1589763: Error changing CD for a running VM when ISO image is on a block domain](https://bugzilla.redhat.com/1589763)
+- [Bug 1868643: Volume not deactivated after ejecting CDROM](https://bugzilla.redhat.com/1868643)
+
+
+### Functionality added with this feature
+
+This feature adds new APIs to vdsm for loading, ejecting and changing CD.
+With these functions, all aforementioned functionality will work properly.
+
+#### Vdsm
+
+- Add VM.insertCD API
+- Add VM.ejectCD API
+- Updated recovery flow to check `in_progress` flag for CDs
+
+
+### Flows
+
+#### VM with empty tray, insert CD
+
+1. engine sends `insertCD` command, specifying ISO by PDIV format
+1. add new drivespec to VM metadata with new attribute `in_progress`
+```
+    <ovirt-vm:device devtype="disk" name="sdc" in_progress="yes">
+```
+1. prepare new drivespec
+  - if failed:
+    1. remove drivespec (log errors)
+    1. fail the flow with original error
+1. loads CD into VM
+  - if failed:
+    1. teardown the volume (log error)
+    1. remove the drive spec (log error)
+    1. fail the flow with the original error
+1. remove `in-progress` attribute from VM metadata
+  - if failed: log the error and succeed
+
+
+#### VM with loaded CD, eject CD
+
+1. engine sends `ejectCD`, specifying ISO path by PDIV format 
+1. vdsm updates VM metadata with new attribute `in_progress`:
+```
+    <ovirt-vm:device devtype="disk" name="sdc" in_progress="yes">
+```
+1. vdsm eject CD from VM (calling libvirt `updateDeviceFlags()`)
+  - if failed (e.g. timeout?):
+    1. remove the `in_progress` flag (log error)
+    1. fail with original error
+1. tear down drivespec
+  - if tearing down volume fails: log error (volume used by another VM)
+1. remove drivespec from VM metadata
+  - if failed: log error and succeed
+
+
+#### VM recovery flow
+
+Try to cleanup storage leftovers from insert and remove operations without changing VM state.
+
+1. metadata in progress, volume is not active
+  - remove metadata from XML
+1. metadata in progress, volume active, no device
+  - deactivate the volume
+  - remove metadata from XML
+1. metadata in progress, volume active, device exists
+  - remove attribute `in_progress` from metadata
```

Comment:
Can there be active asynchronous operations in progress that could do something during or after the cleanup? For example, could it happen that CD removal starts, Vdsm  restarts, identifies the case "metadata in progress, volume active, device exists" and then the CD removal in libvirt finishes?

Reply:
I need to get more familiar with virt vdsm part to be able to answer this, but IIRC handling async livirt events is not properly done in vdsm and this can happen. However, IMHO fixing it is far behind the scope of this feature and if we want to handle it, we need propre async handling of libvirt events.

Reply:
We cannot base anything on libvirt events since we may miss events during
restart.

If a remove operation finish after vdsm recover, the cd will remain removed with
active volume. To handle this we need support from libvirt, like having a block job
so we can detect an operation in progress in libvirt side.

To have this libvirt must support job id set by the caller. We will start a removal
process by storing the job id in the vm metadata. Then if vdsm recover during
the operation we can check the status of the job properly, and register a callback
for completing the cleanup when the job completes.

Reply:
I don't know how it works in libvirt and whether what I write above can actually happen. But it's probably worth to check and document possible problems if there can be any.

Reply:
Based on https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainUpdateDeviceFlags the call returns 0 in case of success, -1 in case of failure. So my impression (don't know how to find it out for sure without reading the source code) is that this call is synchronous and scenarios outlined above cannot happen.